src/lib/reasoners/bitlist.ml

```diff
@@ -300,3 +300,59 @@ let mul a b =
     in
     concat (unknown (sz - width mid_bits - width low_bits) Ex.empty) @@
     concat mid_bits low_bits
+
+let shl a b =
+  (* If the minimum value for [b] is larger than the bitwidth, the result is
+     zero.
+
+     Otherwise, any low zero bit in [a] is also a zero bit in the result, and
+     the minimum value for [b] also accounts for that many minimum zeros (e.g.
+     ?000 shifted by at least 2 has at least 5 low zeroes).
+
+     NB: we would like to use the lower bound from the interval domain for [b]
+     here. *)
+  match Z.to_int (increase_lower_bound b Z.zero) with
+  | n when n < width a ->
+    let low_zeros = Z.trailing_zeros @@ Z.lognot @@ a.bits_clr in
+    if low_zeros + n >= width a then
+      exact (width a) Z.zero (Ex.union (explanation a) (explanation b))
+    else if low_zeros + n > 0 then
+      concat (unknown (width a - low_zeros - n) Ex.empty) @@
+      exact (low_zeros + n) Z.zero (Ex.union (explanation a) (explanation b))
```

Comment:
I understand we cannot keep bits set to one because we only consider the lower bound of `b`, so we may clear all the bits if the value in `b` is too large. Should we consider the upper bound of `b` to keep bits set to one if possible?
For instance, if the bitlist of `a` is `1?11` and the width of `b`is `1`, we can return the bitlist `?110` instead of `?000`?

Reply:
If `a` is `1?11` and `b` is at most `1`, the possible values of the shift are `1?11` and `?110`, so we cannot do better than `??1?`.

We could consider trying to keep track of runs of `1` in the low bits whose length is at smaller than the maximum value of `b`, but this is fairly complex code to write — we could do it, but I'm not sure the complexity/benefits ratio is good enough.